src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs

```diff
@@ -0,0 +1,646 @@
+                    _dataStream = GetDataStream(archiveStream, copyData);
+                    if (_dataStream is SeekableSubReadStream)
+                    {
+                        TarHelpers.AdvanceStream(archiveStream, _size);
+                    }
+                    else if (_dataStream is SubReadStream)
+                    {
+                        // This stream gives the user the chance to optionally read the data section
+                        // when the underlying archive stream is unseekable
+                        skipBlockAlignmentPadding = false;
+                    }
+
+                    break;
+            }
+
+            if (skipBlockAlignmentPadding)
+            {
+                if (_size > 0)
+                {
+                    TarHelpers.SkipBlockAlignmentPadding(archiveStream, _size);
+                }
+
+                if (archiveStream.CanSeek)
+                {
+                    _endOfHeaderAndDataAndBlockAlignment = archiveStream.Position;
+                }
+            }
+        }
+
+        // Returns a stream that represents the data section of the current header.
+        // If copyData is true, then a total number of _size bytes will be copied to a new MemoryStream, which is then returned.
+        // Otherwise, if the archive stream is seekable, returns a seekable wrapper stream.
+        // Otherwise, it returns an unseekable wrapper stream.
+        private Stream? GetDataStream(Stream archiveStream, bool copyData)
+        {
+            if (_size == 0)
+            {
+                return null;
+            }
+
+            if (copyData)
+            {
+                MemoryStream copiedData = new MemoryStream();
+                TarHelpers.CopyBytes(archiveStream, copiedData, _size);
+                return copiedData;
+            }
+
+            return archiveStream.CanSeek
+                ? new SeekableSubReadStream(archiveStream, archiveStream.Position, _size)
+                : (Stream)new SubReadStream(archiveStream, 0, _size);
+        }
+
+        // Attempts to read the fields shared by all formats and stores them in their expected data type.
+        // Throws if end of stream is reached or if any data type conversion fails.
+        // Returns true on success, false if checksum is zero.
+        private bool TryReadCommonAttributes(Stream archiveStream)
+        {
+            byte[] nameBytes = new byte[FieldLengths.Name];
+            byte[] modeBytes = new byte[FieldLengths.Mode];
+            byte[] uidBytes = new byte[FieldLengths.Uid];
+            byte[] gidBytes = new byte[FieldLengths.Gid];
+            byte[] sizeBytes = new byte[FieldLengths.Size];
+            byte[] mTimeBytes = new byte[FieldLengths.MTime];
+            byte[] checksumBytes = new byte[FieldLengths.Checksum];
+            byte[] typeFlagByte = new byte[1];
+            byte[] linkNameBytes = new byte[FieldLengths.LinkName];
```

Comment:
This looks like a lot of allocation.  Would stack allocation work here?  Array pool?  Using one array to reach each of these in turn rather than allocating for each of them and doing everything essentially in parallel?

Reply:
I changed this to use a single rented byte buffer that retrieves the whole header record in one read, converted all the related methods to use spans instead of accessing the stream directly, and now the data is found by slicing that span buffer.